### ISSUE TITLE
Hutchinson divergence estimate discrepancy

### ISSUE THREAD

**Issue by m-lyon** [facebookresearch/flow_matching#49]
Equation 3.3 in the paper uses the random variable $Z \in \mathbb{R}^{d \times d}$ in the estimator of the log-likelihood, where $d$ is the dimensionality of the data, e.g. $x \in \mathbb{R}^{d}$. In the [ODESolver](https://github.com/facebookresearch/flow_matching/blob/47c439602a360ef7d2ec2b9a521b4dda5f335112/flow_matching/solver/ode_solver.py#L147) however, `z` is the same dimension of the data $d$, not $d \times d$.

Is this a mistake in the code or have I missed some simplification/assumption?

**Comment by rtqichen** [comment]
The paper has a typo and will be corrected. See a simple derivation of the Hutchinson trace estimator here: https://blog.shakirm.com/2015/09/machine-learning-trick-of-the-day-3-hutchinsons-trick/. It relies on using any random variable z of dimension d with E[zz^T] = I and the cyclic property of trace. Thanks for bringing this to our attention.

**Comment by aurelio-amerio** [comment]
I'd like to follow up on this issue regarding the calculation of the expectation value.

It appears that `z` is sampled according to `z = (torch.randn_like(x_1).to(x_1.device) < 0) * 2.0 - 1.0`. This results in `z` having the same size as `x_1`, specifically `(batch_size, dimension)`.

Consequently, when evaluating with a `batch_size` of 1 (i.e., a single value for `x_1`), the expectation value is effectively computed using only one sample of `z`. This single-sample Monte Carlo approximation will likely lead to high variance and potentially inaccurate results.

As a reminder, the error in a Monte Carlo estimate of the mean is proportional to $\sigma / \sqrt{N}$, where $\sigma$ is the standard deviation and $N$ is the number of samples. With $N=1$, the error is simply $\sigma$. Increasing to $N=100$ would reduce the error by a factor of 10, and $N=10,000$ by a factor of 100.

Given this, it's not immediately clear that the computational advantage of the Hutchinson trick for approximating the trace is realized when using only a single sample for the expectation. The benefit seems questionable unless the dimensionality is very low (e.g., significantly less than 100), or multiple samples of `z` are used to reduce the variance.

**Comment by rtqichen** [comment]
One can increase the number of samples by duplicating the state. One sample has been enough for training CNFs in the past and for evaluation, since it's averaged over a validation set, the estimate of log-likelihood over the full validation set has quite low variance.
